snapshots/windows/cimfs.go

```diff
@@ -0,0 +1,232 @@
+// writable layers) are same as standard windows snapshotter scratch layers.
+func (s *cimFSSnapshotter) isCimLayer(ctx context.Context, key string) (bool, error) {
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return false, fmt.Errorf("get snapshot info: %w", err)
+	}
+	snCimPath := getCimLayerPath(s.cimDir, id)
+	if _, err := os.Stat(snCimPath); err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *cimFSSnapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
+	baseUsage, err := s.windowsBaseSnapshotter.Usage(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+
+	ctx, t, err := s.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	defer t.Rollback()
+
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, fmt.Errorf("failed to get snapshot info: %w", err)
+	}
+
+	if ok, err := s.isCimLayer(ctx, key); err != nil {
+		return snapshots.Usage{}, err
+	} else if ok {
+		cimUsage, err := cimfs.GetCimUsage(ctx, getCimLayerPath(s.cimDir, id))
+		if err != nil {
+			return snapshots.Usage{}, err
+		}
+		baseUsage.Size += int64(cimUsage)
+	}
+	return baseUsage, nil
+}
+
+func (s *cimFSSnapshotter) Prepare(ctx context.Context, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	m, err := s.createSnapshot(ctx, snapshots.KindActive, key, parent, opts)
+	if err != nil {
+		return m, err
+	}
+	m[0].Type = "CimFS"
+	return m, nil
+}
+
+func (s *cimFSSnapshotter) View(ctx context.Context, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	m, err := s.createSnapshot(ctx, snapshots.KindView, key, parent, opts)
+	if err != nil {
+		return m, err
+	}
+	m[0].Type = "CimFS"
+	return m, nil
+}
+
+func (s *cimFSSnapshotter) Mounts(ctx context.Context, key string) ([]mount.Mount, error) {
+	mounts, err := s.windowsBaseSnapshotter.Mounts(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	mounts[0].Type = "CimFS"
+	return mounts, nil
+}
+
+func (s *cimFSSnapshotter) Commit(ctx context.Context, name, key string, opts ...snapshots.Opt) error {
+	if !strings.Contains(key, snapshots.UnpackKeyPrefix) {
+		return fmt.Errorf("committing a scratch snapshot to read-only cim layer isn't supported yet")
+	}
+
+	return s.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		// grab the existing id
+		id, _, _, err := storage.GetInfo(ctx, key)
+		if err != nil {
+			return fmt.Errorf("failed to get storage info for %s: %w", key, err)
+		}
+		path := s.getSnapshotDir(id)
+
+		usage, err := fs.DiskUsage(ctx, path)
+		if err != nil {
+			return fmt.Errorf("failed to collect disk usage of snapshot storage: %s: %w", path, err)
+		}
+
+		cimUsage, err := cimfs.GetCimUsage(ctx, getCimLayerPath(s.cimDir, id))
+		if err != nil {
+			return err
+		}
+		usage.Size += int64(cimUsage)
```

Comment:
Is there any difference between this and calling `Usage`? Maybe we can reduce duplication?